src/main/scala/chisel3/util/experimental/pla.scala

```diff
@@ -0,0 +1,123 @@
+    * @param invert A [[BitPat]] specify which bit of the output should be inverted. A `1` means the correspond position
+    *               of the output should be inverted in the PLA, a `0` or a `?` means direct output from the OR matrix.
+    * @return the input [[Wire]] of [[UInt]] of the constructed pla.
+    * @return the output [[Wire]] of [[UInt]] of the constructed pla.
+    * @example A 1-of-8 decoder (like the 74xx138) can be constructed as follow
+    *          {{{
+    *            val (inputs, outputs) = pla(Seq(
+    *              (BitPat("b000"), BitPat("b00000001")),
+    *              (BitPat("b001"), BitPat("b00000010")),
+    *              (BitPat("b010"), BitPat("b00000100")),
+    *              (BitPat("b011"), BitPat("b00001000")),
+    *              (BitPat("b100"), BitPat("b00010000")),
+    *              (BitPat("b101"), BitPat("b00100000")),
+    *              (BitPat("b110"), BitPat("b01000000")),
+    *              (BitPat("b111"), BitPat("b10000000")),
+    *            ))
+    *          }}}
+    */
+  def apply(table: Seq[(BitPat, BitPat)], invert: BitPat = BitPat("b0")): (UInt, UInt) = {
+    require(table.nonEmpty, "pla table must not be empty")
+
+    val (inputTerms, outputTerms) = table.unzip
+    require(
+      if (inputTerms.length > 1)
+        inputTerms.tail.map(_.getWidth == inputTerms.head.getWidth).reduce(_ && _)
+      else
+        true,
+      "all `BitPat`s in the input part of specified PLA table must have the same width"
+    )
+    require(
+      if (outputTerms.length > 1)
+        outputTerms.tail.map(_.getWidth == outputTerms.head.getWidth).reduce(_ && _)
+      else
+        true,
+      "all `BitPat`s in the output part of specified PLA table must have the same width"
+    )
+
+    // now all inputs / outputs have the same width
+    val numberOfInputs = inputTerms.head.getWidth
+    val numberOfOutputs = outputTerms.head.getWidth
+
+    val inverterMask = invert.value & invert.mask
+    if (inverterMask.bitCount != 0)
+      require(invert.getWidth == numberOfOutputs,
+        "non-zero inverter mask must have the same width as the output part of specified PLA table"
+      )
+
+    // input wires of the generated PLA
+    val inputs = Wire(UInt(numberOfInputs.W))
+    val invInputs = ~inputs
+
+    // output wires of the generated PLA
+    val outputs = Wire(UInt(numberOfOutputs.W))
+
+    // the AND matrix
+    // use `term -> AND line` map to reuse AND matrix output lines
+    val andMatrixOutputs: Map[String, Bool] = inputTerms.map { t =>
+      val andLine = Cat(
+        Seq
+          .tabulate(numberOfInputs) { i =>
+            if (t.mask.testBit(i)) {
+              Some(
+                if (t.value.testBit(i)) inputs(i)
+                else invInputs(i)
+              )
+            } else {
+              None
+            }
+          }
+          .flatten
+      ).andR()
+      t.toString -> andLine
```

Comment:
The `andLine` can be inlined after the `->`.

Reply:
Sure :)